src/util/LogSlowExecution.cpp

```diff
@@ -44,36 +44,50 @@ LogSlowExecution::checkElapsedTime() const
     {
         std::lock_guard<std::mutex> guard(gLogSlowExecMutex);
 
-        // Only emit a new log message once-per-second (optionally preceded
-        // by a summary of how many messages have been dropped since last time).
-        if ((finish - gLastLogMessage) > std::chrono::seconds(1))
+        // Check whether we're 10 times over the threshold to decide the log
+        // level.
+        LogLevel ll = elapsed > mThreshold * 10 ? LogLevel::LVL_WARNING
+                                                : LogLevel::LVL_DEBUG;
+
+        // Compare the log level of this message and the current log level.
+        if (ll <= Logging::getLogLevel("Perf"))
         {
-            if (gDroppedLogMessagesSinceLast > 0)
-            {
-                CLOG_WARNING(Perf, "Dropped {} slow-execution warning messages",
-                             gDroppedLogMessagesSinceLast);
-                gDroppedLogMessagesSinceLast = 0;
-            }
-            auto msg = fmt::format(FMT_STRING("'{}' {} {} s"), mName, mMessage,
-                                   static_cast<float>(elapsed.count()) / 1000);
-            // if we're 10 times over the threshold, log with INFO, otherwise
-            // use DEBUG
-            if (elapsed > mThreshold * 10)
+            // Only emit a new log message once-per-second (optionally preceded
+            // by a summary of how many messages have been dropped since last
+            // time).
+            if ((finish - gLastLogMessage) > std::chrono::seconds(1))
             {
-                CLOG_WARNING(Perf, "{}", msg);
+                if (gDroppedLogMessagesSinceLast > 0)
```

Comment:
I think this implementation still suffers from the same issue than master: if a very slow execution is sandwiched between "regularly slow" ones, the actual entry will not be logged because things are logged with `CLOG_DEBUG`

Reply:
In your example, is the user expected to _not_ see any messages about "regularly slow" ones (because of the logging level), but is the user expected to see a message about the very slow execution? If that's the case, I believe that this change addresses that.

https://github.com/stellar/stellar-core/blob/89a6bffcd94fc6f48eeb0848f6bc4d3f6328ba9e/src/util/LogSlowExecution.cpp#L53

If the current logging level prevents `CLOG_DEBUG` from actually being printed, then this method will not update `gLastLogMessage` or `gDroppedLogMessagesSinceLast`. `gLastLogMessage` is always the last time we actually printed something and the user saw it, not the time when we called `CLOG_*`. So, when a very slow execution (more than 10x) is sandwiched between slow ones (more than the threshold, less than 10x), then those slow ones won't make it past `if (ll <= Logging::getLogLevel("Perf"))`, and the very slow execution will get logged and the user will see it.

Please let me know if I'm misunderstanding something! @MonsieurNicolas

Reply:
yeah this version is slightly better than master in that at `WARNING` level, it will only count logging events that correspond to `CLOG_WARNING`, *but* at `DEBUG` level can hide the `CLOG_WARNING` sandwiched like I wrote earlier

Reply:
What would be the expected outcome when we have "regularly slow" -> "extremely slow" -> "regularly slow" in succession at the `DEBUG` level?

Would we want the user to see both the first "regularly slow" one _and_ the "extremely slow" one? I was under the impression that we'd only show the first "regularly slow" one.

Maybe what you're saying is that we should keep track of `gLastLogMessage` for `WARNING` (=extremely slow) and `DEBUG` (=regularly slow) separately? That way, we'd decide whether to print `WARNING` messages (=extremely slow) only based on when the last `WARNING` message was, and it wouldn't depend on any `DEBUG` messages. Is my understanding of this correct? @MonsieurNicolas

Reply:
@MonsieurNicolas I updated the code so it keeps track of the last message for each logging level.

If there's a series of events "slow" -> "slow" -> "extremely slow" -> (2 seconds of nothing) -> "slow", then

- If the log level is `DEBUG`
    - The first "slow" is printed.
    - The second "slow" is ignored. (because we just printed a `DEBUG` message)
    - The "extremely slow" event is printed along with `dropped 1 message`
    - The last "slow" is printed.
- If the log level is `WARNING`
    - Only the "extremely slow" event is printed.
    - No message about dropping messages.

Reply:
Would probably be better to just bundle `gLastLogMessage` and `gDroppedLogMessagesSinceLast` ? The map could be a map of `pair<std::chrono::system_clock::time_point, size_t>` ? That way the number of dropped messages is related to what is being logged (and we can fix the log level by logging a single line) and do something like:
```
            std::string skipped;
            if (droppedLogMessagesSinceLast > 0)
            {
                skipped = fmt::format(FMT_STRING(" [skipped: {}]"), droppedLogMessagesSinceLast);
                droppedLogMessagesSinceLast = 0;
            }
            auto msg = fmt::format(FMT_STRING("'{}' {} {} s{}"), mName, mMessage,
                                   static_cast<float>(elapsed.count()) / 1000, droppedLogMessagesSinceLast, skipped);
```

Reply:
Makes sense! I updated the code.